prow/plugins/branchcleaner/branchcleaner.go

```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package branchcleaner
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	pluginName = "branchcleaner"
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	return &pluginhelp.PluginHelp{
+		Description: "The branchcleaner automatically deletes the source branch of Pull Requests if it is on the same repo as the target branch. This is helpful to keep repos that don't allow forking clean."}, nil
+}
+
+func handlePullRequest(pc plugins.Agent, pre github.PullRequestEvent) error {
+	return handle(pc.GitHubClient, pc.Logger, pre)
+}
+
+type githubClient interface {
+	DeleteRef(owner, repo, ref string) error
+}
+
+func handle(gc githubClient, log *logrus.Entry, pre github.PullRequestEvent) error {
+	// Only consider closed PRs
+	if pre.Action != github.PullRequestActionClosed {
+		return nil
+	}
+
+	pr := pre.PullRequest
+	// Only consider merged PRs that were closed via a merge
+	if !pr.Merged || pr.MergeSHA == nil || pr.Base.Repo.DefaultBranch != pr.Base.Ref {
```

Comment:
Why do we check that `MergeSHA` is `nil`? I'm still [not sure when it's empty](https://developer.github.com/v3/pulls/#response-1) (I think it only empty when the PR is not mergable).

Also, can we update the comment on this to explain what's happening? I'm not sure what `pr.Base.Repo.DefaultBranch != pr.Base.Ref` is supposed to check?

Reply:
Good point, I used the [`updateconfig`](https://github.com/kubernetes/test-infra/blob/7162eefc42681f6192cab42e4031e6893e920cc2/prow/plugins/updateconfig/updateconfig.go#L136) as a template for this but yeah, we probably want to clean up even when the PR does not go into the default branch